From e1004db899d5b7974b25ab667a1b55c5de7b80fc Mon Sep 17 00:00:00 2001
From: George Moussalem <george.moussalem@outlook.com>
Date: Wed, 05 Aug 2026 12:20:15 +0400
Subject: [PATCH v2 11/17] dt-bindings: usb: qcom,snps-dwc3: Add support for USB phy mux selection
MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 7bit
Message-Id: <20260805-ipq5018-usb3-v2-11-eabd65f41443@outlook.com>

On some SoCs, such as IPQ5018 and IPQ5332, the USB3 PHY and PCIe PHY
pads are muxed to the same physical pads. Selection is controlled by a
register in TCSR. The mux register must be written to before the USB
controller is taken out of reset and clocks are enabled.
As such, add a property to enable USB mux selection that takes a phandle
to the TCSR syscon and the offset to the mux register.

Signed-off-by: George Moussalem <george.moussalem@outlook.com>
Signed-off-by: Stanislaw Pal <kuncy7@gmail.com>
---
Credits to Stanislaw Pal for finding the right order of applying the USB
mux selection ahead of taking the USB controller out of reset and
enabling clocks.
---
 Documentation/devicetree/bindings/usb/qcom,snps-dwc3.yaml | 12 ++++++++++++
 1 file changed, 12 insertions(+)

--- a/Documentation/devicetree/bindings/usb/qcom,snps-dwc3.yaml
+++ b/Documentation/devicetree/bindings/usb/qcom,snps-dwc3.yaml
@@ -140,6 +140,16 @@ properties:
     minItems: 3
     maxItems: 19
 
+  qcom,phy-usb-mux-sel:
+    description: |
+      Mux selection for USB PHY. On some SoCs, the USB3 and PCIe PHY pads are
+      muxed and the selection is done via a register in TCSR syscon.
+    $ref: /schemas/types.yaml#/definitions/phandle-array
+    items:
+      - items:
+          - description: phandle of TCSR syscon
+          - description: offset of PHY mux selection register
+
   qcom,select-utmi-as-pipe-clk:
     description:
       If present, disable USB3 pipe_clk requirement.
@@ -469,6 +479,8 @@ allOf:
             - const: pwr_event
             - const: dp_hs_phy_irq
             - const: dm_hs_phy_irq
+      required:
+        - qcom,phy-usb-mux-sel
 
   - if:
       properties:
